src/Desktop/UIAutomation/A11yAutomation.cs

```diff
@@ -46,6 +45,63 @@ public static DesktopElement ElementFromProcessId(int pid)
             return elements != null ? elements.First() : null;
         }
 
+        private static bool FindProcessMatchingChildren(IUIAutomationElement parent, IUIAutomationTreeWalker walker, int pid, IList<IUIAutomationElement> matchingElements, IList<IUIAutomationElement> nonMatchingElements)
+        {
+            for (var child = walker.GetFirstChildElement(parent); child != null; child = walker.GetNextSiblingElement(child))
+            {
+                if (child.CurrentProcessId == pid)
+                {
+                    matchingElements.Add(child);
+                }
+                else
+                {
+                    nonMatchingElements.Add(child);
+                }
+            }
+
+            return matchingElements.Any();
+        }
+
+        private static IList<IUIAutomationElement> FindProcessMatchingChildrenOrGrandchildren(IUIAutomationElement root, int pid)
+        {
+            IUIAutomationTreeWalker walker = GetTreeWalker(TreeViewMode.Control);
+            List<IUIAutomationElement> matchingElements = new List<IUIAutomationElement>();
+            List<IUIAutomationElement> nonMatchingElements = new List<IUIAutomationElement>();
+            List<IUIAutomationElement> nonMatchingElementsSecondLevel = new List<IUIAutomationElement>();
+
+            if (FindProcessMatchingChildren(root, walker, pid, matchingElements, nonMatchingElements))
+            {
+                CleanUpNonMatchingElements();
+                return matchingElements;
+            }
+
+            foreach (var nonMatchingElement in nonMatchingElements)
+            {
+                if (FindProcessMatchingChildren(nonMatchingElement, walker, pid, matchingElements, nonMatchingElementsSecondLevel))
+                {
+                    CleanUpNonMatchingElements();
+                    return matchingElements;
+                }
+            }
+            CleanUpNonMatchingElements();
+            return matchingElements; // Will always be empty
+
+            void CleanUpNonMatchingElements()
+            {
+                foreach (var nonMatchingElementToBeCleaned in nonMatchingElements)
+                {
+                    Marshal.ReleaseComObject(nonMatchingElementToBeCleaned);
+                }
+                nonMatchingElements = null;
+                foreach (var nonMatchingElementToBeCleaned in nonMatchingElementsSecondLevel)
+                {
+                    Marshal.ReleaseComObject(nonMatchingElementToBeCleaned);
+                }
+                nonMatchingElementsSecondLevel = null;
```

Comment:
There's probably no need to set the lists to null, especially if we drop the nested function